src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -195,31 +206,34 @@ public MemoryPoolIterator ConsumingStart()
                 }
                 else
                 {
+                    // Dispose won't have returned the blocks if we were consuming, so return them now
                     returnStart = _head;
-                    returnEnd = null;
                     _head = null;
                     _tail = null;
                 }
 
-                ReturnBlocks(returnStart, returnEnd);
-
-                if (!_consuming)
```

Comment:
It looks like github lost my comment earlier where I realized that all access to `_consuming` really does need to be guarded by the lock because of the interaction with Dispose.